packages-exp/auth-exp/src/core/errors.ts

```diff
@@ -0,0 +1,332 @@
+  INVALID_CODE = 'invalid-verification-code',
+  INVALID_CONTINUE_URI = 'invalid-continue-uri',
+  INVALID_CORDOVA_CONFIGURATION = 'invalid-cordova-configuration',
+  INVALID_CUSTOM_TOKEN = 'invalid-custom-token',
+  INVALID_DYNAMIC_LINK_DOMAIN = 'invalid-dynamic-link-domain',
+  INVALID_EMAIL = 'invalid-email',
+  INVALID_IDP_RESPONSE = 'invalid-credential',
+  INVALID_MESSAGE_PAYLOAD = 'invalid-message-payload',
+  INVALID_OAUTH_CLIENT_ID = 'invalid-oauth-client-id',
+  INVALID_OAUTH_PROVIDER = 'invalid-oauth-provider',
+  INVALID_OOB_CODE = 'invalid-action-code',
+  INVALID_ORIGIN = 'unauthorized-domain',
+  INVALID_PASSWORD = 'wrong-password',
+  INVALID_PERSISTENCE = 'invalid-persistence-type',
+  INVALID_PHONE_NUMBER = 'invalid-phone-number',
+  INVALID_PROVIDER_ID = 'invalid-provider-id',
+  INVALID_RECIPIENT_EMAIL = 'invalid-recipient-email',
+  INVALID_SENDER = 'invalid-sender',
+  INVALID_SESSION_INFO = 'invalid-verification-id',
+  INVALID_TENANT_ID = 'invalid-tenant-id',
+  MISSING_ANDROID_PACKAGE_NAME = 'missing-android-pkg-name',
+  MISSING_APP_CREDENTIAL = 'missing-app-credential',
+  MISSING_AUTH_DOMAIN = 'auth-domain-config-required',
+  MISSING_CODE = 'missing-verification-code',
+  MISSING_CONTINUE_URI = 'missing-continue-uri',
+  MISSING_IFRAME_START = 'missing-iframe-start',
+  MISSING_IOS_BUNDLE_ID = 'missing-ios-bundle-id',
+  MISSING_OR_INVALID_NONCE = 'missing-or-invalid-nonce',
+  MISSING_PHONE_NUMBER = 'missing-phone-number',
+  MISSING_SESSION_INFO = 'missing-verification-id',
+  MODULE_DESTROYED = 'app-deleted',
+  MFA_REQUIRED = 'multi-factor-auth-required',
+  NEED_CONFIRMATION = 'account-exists-with-different-credential',
+  NETWORK_REQUEST_FAILED = 'network-request-failed',
+  NULL_USER = 'null-user',
+  NO_AUTH_EVENT = 'no-auth-event',
+  NO_SUCH_PROVIDER = 'no-such-provider',
+  OPERATION_NOT_ALLOWED = 'operation-not-allowed',
+  OPERATION_NOT_SUPPORTED = 'operation-not-supported-in-this-environment',
+  POPUP_BLOCKED = 'popup-blocked',
+  POPUP_CLOSED_BY_USER = 'popup-closed-by-user',
+  PROVIDER_ALREADY_LINKED = 'provider-already-linked',
+  QUOTA_EXCEEDED = 'quota-exceeded',
+  REDIRECT_CANCELLED_BY_USER = 'redirect-cancelled-by-user',
+  REDIRECT_OPERATION_PENDING = 'redirect-operation-pending',
+  REJECTED_CREDENTIAL = 'rejected-credential',
+  TENANT_ID_MISMATCH = 'tenant-id-mismatch',
+  TIMEOUT = 'timeout',
+  TOKEN_EXPIRED = 'user-token-expired',
+  TOO_MANY_ATTEMPTS_TRY_LATER = 'too-many-requests',
+  UNAUTHORIZED_DOMAIN = 'unauthorized-continue-uri',
+  UNSUPPORTED_PERSISTENCE = 'unsupported-persistence-type',
+  UNSUPPORTED_TENANT_OPERATION = 'unsupported-tenant-operation',
+  USER_CANCELLED = 'user-cancelled',
+  USER_DELETED = 'user-not-found',
+  USER_DISABLED = 'user-disabled',
+  USER_MISMATCH = 'user-mismatch',
+  USER_SIGNED_OUT = 'user-signed-out',
+  WEAK_PASSWORD = 'weak-password',
+  WEB_STORAGE_UNSUPPORTED = 'web-storage-unsupported'
+}
+
+const ERRORS: ErrorMap<AuthErrorCode> = {
```

Comment:
I'll start the discussion here - is there a way we can make this optional? All these strings will be a large part of the compiled binary

Reply:
As discussed offline, it might be good to have separate error maps for "dev" and "prod", that way the user could choose which gets included. That would require the AUTH_ERROR_FACTORY to not be a constant.

I think we can mull this for a while, but to make that change more straightforward in the future, maybe this module should export a function instead of a constant?

export function getErrorFactory() { return AUTH_ERROR_FACTORY }

Reply:
interested in what others think about... @bojeil-google ?

Reply:
Just as a FYI, Firestore implemented something similar recently, but in a different way. They split their asserts into hardAsserts and debugAsserts, and debugAsserts are removed in the prod build.

https://github.com/firebase/firebase-js-sdk/pull/2859
https://github.com/firebase/firebase-js-sdk/pull/2814

Reply:
The assert model is cool, but doesn't work great in auth since most of our error messages are translations for server side errors

Reply:
Note that this list also includes client side errors that do not map to server side errors.
So you are suggesting to make error mapping functionality modular? So if the developer does not import it, then what happens?
Expecting the developer to know what errors to include puts a lot of burden on the developer. They would need to understand each error.

Reply:
I think for production builds we could just have a short code that can be mapped in our documentation, something like Error: auth/email-exists, with the full description & troubleshooting tips available online

Reply:
this would be a breaking API change though, and we would need the page up on the documentation before we could launch this, but it would be cool and could potentially make the final binary that much smaller

Reply:
I think I misunderstood your suggestion. You are suggesting modularization of the error messages population component and not the entire client side errors (and how they are mapped from server to client). That may not be terrible but a breaking change, alas.

I agree we should optimize this but I would suggest we think about it more (this also seems like a general problem and not specific to Auth). JsCore should try to find a unified solution for it. I would recommend implementing the current spec in its current form for now and then experimenting with the generated binary and revisiting this to see how we can avoid bundling these messages.

Reply:
agree on punting this to JS core for now, lets proceed as is, we can measure the possible effect of the strings during alpha and maybe implement this during beta